services/src/main/java/org/keycloak/protocol/oidc/endpoints/TokenRevocationEndpoint.java

```diff
@@ -0,0 +1,211 @@
+
+        if (client.isBearerOnly()) {
+            throw new CorsErrorResponseException(cors, OAuthErrorException.INVALID_CLIENT, "Bearer-only not allowed",
+                Response.Status.BAD_REQUEST);
+        }
+    }
+
+    private void checkToken() {
+        String encodedToken = formParams.getFirst(PARAM_TOKEN);
+
+        if (encodedToken == null) {
+            event.error(Errors.INVALID_REQUEST);
+            throw new CorsErrorResponseException(cors, OAuthErrorException.INVALID_REQUEST, "Token not provided",
+                Response.Status.BAD_REQUEST);
+        }
+
+        token = session.tokens().decode(encodedToken, RefreshToken.class);
+
+        if (token == null) {
+            event.error(Errors.INVALID_TOKEN);
+            throw new CorsErrorResponseException(cors, OAuthErrorException.INVALID_TOKEN, "Invalid token", Response.Status.OK);
+        }
+
+        if (!(TokenUtil.TOKEN_TYPE_REFRESH.equals(token.getType()) || TokenUtil.TOKEN_TYPE_OFFLINE.equals(token.getType()))) {
+            event.error(Errors.INVALID_TOKEN_TYPE);
+            throw new CorsErrorResponseException(cors, OAuthErrorException.UNSUPPORTED_TOKEN_TYPE, "Unsupported token type",
+                Response.Status.BAD_REQUEST);
+        }
+    }
+
+    private void checkIssuedFor() {
+        String issuedFor = token.getIssuedFor();
+        if (issuedFor == null) {
+            event.error(Errors.INVALID_TOKEN);
+            throw new CorsErrorResponseException(cors, OAuthErrorException.INVALID_TOKEN, "Invalid token", Response.Status.OK);
+        }
+
+        if (!client.getClientId().equals(issuedFor)) {
+            event.error(Errors.INVALID_REQUEST);
+            throw new CorsErrorResponseException(cors, OAuthErrorException.INVALID_REQUEST, "Unmatching clients",
+                Response.Status.BAD_REQUEST);
+        }
+    }
+
+    private void checkUser() {
+        UserSessionModel userSession = new UserSessionCrossDCManager(session).getUserSessionWithClient(realm,
+            token.getSessionState(), false, client.getId());
+
+        if (userSession == null) {
+            userSession = new UserSessionCrossDCManager(session).getUserSessionWithClient(realm, token.getSessionState(), true,
+                client.getId());
+
+            if (userSession == null) {
+                event.error(Errors.USER_SESSION_NOT_FOUND);
+                throw new CorsErrorResponseException(cors, OAuthErrorException.INVALID_TOKEN, "Invalid token",
+                    Response.Status.OK);
+            }
+        }
+
+        user = userSession.getUser();
+
+        if (user == null) {
+            event.error(Errors.USER_NOT_FOUND);
+            throw new CorsErrorResponseException(cors, OAuthErrorException.INVALID_TOKEN, "Invalid token", Response.Status.OK);
+        }
+
+        event.user(user);
+    }
+
+    private void revokeClient() {
+        session.users().revokeConsentForClient(realm, user.getId(), client.getId());
```

Comment:
Currently, consents are persistent and we don't ask end-users for consent if they have previously granted it to a client.

By removing this you are forcing users to consent again. Is that on purpose?

Reply:
@pedroigor 
> By removing this you are forcing users to consent again. Is that on purpose?

Yes. A user revokes a token because the user doesn't want the client to use it, so the relationship between the user and the client should be broken off and also the consent should be removed.

Reply:
The user does not revoke tokens but clients. Another reason behind my question is that reading the specs it is not clear how consents should be handled when revoking tokens. It does talk about cleaning up data associated with the token (e.g.: session data).

IMO, consents span multiple sessions and are not related to a single access token. Users have control over their consent so they can decide whether or not they want to revoke consent for a given client. By revoking consent when revoking tokens you are assuming the end-user also wants to revoke the consent but is that always true?

I do see a lot of references where token revocation is more a "logout"  feature for clients than a revocation mechanism for consents.